### PR TITLE
INTG-600 user management API doc

### PR DESCRIPTION
**Pull request by mrinalirao**
user management API doc

### CODE REVIEW

File: source/index.html.md [hunk 1]
```diff
@@ -2639,3 +2639,149 @@ Status code     | Description
...
+`user_id`         | The ID of the user
+
+### Response
+
+Status code     | Description
+-----------     | -----------
+200 OK          | Deletion successful
+400 Bad Request | Invalid request 
+403 Forbidden   | No permission to delete the user
+404 Not Found   | The group ID/ organisation ID or the user ID specified was not found
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: What if the user not found? What if 404 from internal services?

File: source/index.html.md [hunk 2]
```diff
@@ -2639,3 +2639,149 @@ Status code     | Description
...
+The body should use the content type `application/json` with the following parameters:
+
+Parameter          | Description
+---------          | -----------
+`users`            | The array of user accounts to add. See table below for allowed parameters for a user account.
+
+User Accounts      | Description
+---------          | -----------
+`firstName`        | FirstName of the user (with maximum character limit of 150)
+`LastName`         | LastName of the user (with maximum character limit of 150)
```
[the remark this answers is not in the document]
Comment (mrinalirao) [reply source/index.html.md]: Yes the same as mentioned above (150 for names and 200 for email), this is as per mullet

File: source/index.html.md [hunk 3]
```diff
@@ -2639,3 +2639,149 @@ Status code     | Description
...
+`email`            | Email Address of the user (with maximum character limit of 200)
+`password`         | Password of the user (with maximum character limit of 200 and minimum character limit of 6) (optional)
```
[the remark this answers is not in the document]
Comment (mrinalirao) [reply source/index.html.md]: nothing actually happens...the user gets an email and a link which takes him to his account directly

File: source/index.html.md [hunk 4]
```diff
@@ -2639,3 +2639,149 @@ Status code     | Description
...
+---------          | -----------
+`firstName`        | FirstName of the user (with maximum character limit of 150)
+`LastName`         | LastName of the user (with maximum character limit of 150)
+`email`            | Email Address of the user (with maximum character limit of 200)
+
+### Response
+
+Status code     | Description
+-----------     | -----------
+200 OK          | Users added
```
[the remark this answers is not in the document]
Comment (mrinalirao) [reply source/index.html.md]: This does not indicate how many users were added, this indicates that a request was sent and the users are being processed..number of users added/failed will only come in via email once the entire queue has been processed. Let me know whether i need to change the description here

File: source/index.html.md (hunk 1 above)
Comment (mrinalirao) [reply source/index.html.md]: it sends the same error to the pub : 
`return err.isBoom ? err : toBoom(err);`

File: source/index.html.md [hunk 5]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+See the [Errors](#error-codes) section for more details.
+
+## Remove User
+
+<aside class="warning">
+User management endpoints are released in Beta and may change depending on feedback during trial.
+</aside>
+
+Enable administrators (users with the "Manage Role") to remove a user from the organization of the requesting user.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: "Manage Role" permission

Comment (dimitrikoubaroulis) [reply source/index.html.md]: to remove a user from a specified group or organisation

File: source/index.html.md [hunk 6]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+<aside class="notice">
+To remove a user from an organisation, the following conditions must be satisfied:
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: from a user group

File: source/index.html.md [hunk 7]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+1. The requesting user cannot remove themselves
+2. Only users with the Manage Role permission can remove other users
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: `Manage Role`

File: source/index.html.md [hunk 8]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+3. Admin users (users with "manage_role" permissions) can remove other administrator users from the organisation
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Admin users -> Administrators

File: source/index.html.md [hunk 9]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+4. If a user is added to an administrator group (group with "manage_role" permissions) within the organisation and if the user to be removed is the only member of a group, then a check is made to see if the organisation has more than one administrator, if yes, the user can be deleted from the group.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: manage_role => `Manage Role`

Comment (dimitrikoubaroulis) [inline source/index.html.md]: This sentence does not make sense to me. Consider clarifying please. Also please choose 'deleted' or 'removed' and use it consistently in doco.

File: source/index.html.md [hunk 10]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+5. If a group has administrator privileges, then all the users in the group inherit the administrator privileges, removing a user from this group will cause the user to lose his administrator privilges (unless the user was already an administrator of the organisation before being added to the group)
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: What is administrator privileges? How does a group get them? Also this sentence needs breaking up. It does not make grammatical sense as it is.

File: source/index.html.md [hunk 11]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+6. The last administrator of the organisation cannot be removed
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: The last remaining

File: source/index.html.md [hunk 12]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+7. If the user is an inactive member of the group, the user can be removed
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: This is not a condition that must be satisfied (as you say at the top of this list). Do we need this line at all?

File: source/index.html.md [hunk 13]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+To lookup the user ID with the [Search Users](#search-users) endpoint
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: with -> use ?   To lookup the user ID required by the remove user endpoint use the ....

Comment (dimitrikoubaroulis) [reply source/index.html.md]: To lookup the group ID of required by remove user endpoint use the ...

File: source/index.html.md [hunk 14]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+### HTTP Request
+
+`DELETE https://api.safetyculture.io/groups/<group_id>/users/<user_id>`
+
+### URL Parameters
+
+Parameter         | Description
+---------         | -----------
+`group_id`        | The ID of the organisation or a group within the organisation to remove the user from
+`user_id`         | The ID of the user
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: of the user to remove

File: source/index.html.md (hunk 1 above)
Comment (dimitrikoubaroulis) [reply source/index.html.md]: 404 could be returned for a number of reasons e.g. we forgot to add an internal service endpoint. This is a general problem we have with our error handling overall. Can chat about it sometime.

File: source/index.html.md [hunk 15]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+User Accounts      | Description
+---------          | -----------
+`firstName`        | The first name of the user (with maximum character limit of 150)
+`lastName`         | The last name of the user (with maximum character limit of 150)
+`email`            | Email Address of the user (with maximum character limit of 200)
+`password`         | Password of the user (with maximum character limit of 200 and minimum character limit of 6) (optional)
+`hashedPassword`   | Hashed password of the user (with maximum character limit of 60) (optional)
+
+
+The responses should have the following properties:
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: responses -> user entries in the response payload have the following properties

File: source/index.html.md [hunk 16]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+`POST https://api.safetyculture.io/users`
+
+### Request Body
+
+The body should use the content type `application/json` with the following parameters:
+
+Parameter          | Description
+---------          | -----------
+`users`            | The array of user accounts to add. See table below for allowed parameters for a user account.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: parameters -> properties
for a user account -> to initialise a user account with

File: source/index.html.md [hunk 17]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+Enable administrators (users with the "Manage Role") to add and remove users to and from the organization of the requesting user.
+
+## Add Users
+
+<aside class="warning">
+User management endpoints are released in Beta and may change depending on feedback during trial.
+</aside>
+
+The Add Users endpoint allows an administrator (user with "manage_role" permissions) to add a user to his organisation.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: `Add Users` (with backticks)
`Manage Role` (with backticks) 
permissions -> permission
his -> their

File: source/index.html.md [hunk 18]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+This depends on the subscription(number of user seats) of the organisation.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: space before parenthesis

File: source/index.html.md [hunk 19]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+If the organisation has already exhausted it's maximum number of user seats, a 400 Error will be thrown with the message: `Not enough user seats to invite new user`
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: it's -> its
400 -> 422 unprocessable entity may be more appropriate here
thrown -> returned
No need to include the actual error message here as we may change it
You have not talked about invitations until now btw. Are you going to cover this part in the doco? That a user gets invited when added?

File: source/index.html.md [hunk 20]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+
+After all the users have been processed, the administrator will be notified via an email on how many users were added successfully to the organisation and how many failed.
+
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: A maximum of 500 users can be added at one time in a single request. Can we remove this limit and spawn multiple mullet requests inside thepub. Just idea, please take a note to discuss when we have time.

File: source/index.html.md [hunk 21]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+A user is notified via email when added to an organisation, if the user was added without a password (since providing a password is optional), 
+the email contains a link which the user can click to reset the password.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Split this sentence here into two. The email sent to the added user, contains a link inviting the user to sign in to iAuditor. If a password was submitted in the `Add Users` request for this user, the user will be taken to the iAuditor login screen to sign in with that password. If no password was submitted, the user will be taken to an iAuditor screen to set a new password. Is that correct? Please clarify.

File: source/index.html.md [hunk 22]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+User management endpoints are released in Beta and may change depending on feedback during trial.
+</aside>
+
+The Add Users endpoint allows an administrator (user with "manage_role" permissions) to add a user to his organisation.
+This depends on the subscription(number of user seats) of the organisation.
+If the organisation has already exhausted it's maximum number of user seats, a 400 Error will be thrown with the message: `Not enough user seats to invite new user`
+
+After all the users have been processed, the administrator will be notified via an email on how many users were added successfully to the organisation and how many failed.
+
+A user is notified via email when added to an organisation, if the user was added without a password (since providing a password is optional),
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Upon successful addition of a user, that user receives an email notifying them and inviting them to access iAuditor in their default web browser.

File: source/index.html.md [hunk 23]
```diff
@@ -2639,3 +2639,158 @@ Status code     | Description
...
+---------          | -----------
+`users`            | The array of user accounts to add. See table below for allowed parameters for a user account.
+
+User Accounts      | Description
+---------          | -----------
+`firstName`        | The first name of the user (with maximum character limit of 150)
+`lastName`         | The last name of the user (with maximum character limit of 150)
+`email`            | Email Address of the user (with maximum character limit of 200)
+`password`         | Password of the user (with maximum character limit of 200 and minimum character limit of 6) (optional)
+`hashedPassword`   | Hashed password of the user (with maximum character limit of 60) (optional)
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: In the one above, the (optional) is on the left column. Here it is at the end. Consistency please :-)

Comment (dimitrikoubaroulis) [reply source/index.html.md]: Please add a couple of lines about what is hashed Password and why user should give it

File: source/index.html.md (hunk 19 above)
Comment (mrinalirao) [reply source/index.html.md]: the 400 is returned from mullet..should i change this in mullet? a 400 error can come from a lot of services in the chain, mullet, soter, kickback etc..shud i wrap this specific error in the pub to explicitly return 422?

File: source/index.html.md (hunk 23 above)
Comment (mrinalirao) [reply source/index.html.md]: did not understand the first comment :) Need more information from tristan regarding the hashed password...will update this shortly

File: source/index.html.md [hunk 24]
```diff
@@ -2639,3 +2639,163 @@ Status code     | Description
...
+</aside>
+
+Enable administrators (users with the `Manage Role` permission) to remove a user from a specified group or organisation
+<aside class="notice">
+To remove a user from a user group, the following conditions must be satisfied:
+
+1. The requesting user cannot remove themselves
+2. Only users with the `Manage Role` permission can remove other users
+3. Administrators (users with `Manage Role` permission) can remove other administrator users from the organisation
+4. If a user is added to an administrator group (group with `Manage Role` permission) within the organisation, the user gets the administrator (`Manage Role` permission) privileges,
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: If the user to be removed belongs to a group with `Manage Role` permission they inherit the `Manage Role` permission. In that case the user can only be removed from the group if they are not the only administrator in the organisation.

File: source/index.html.md [hunk 25]
```diff
@@ -2639,3 +2639,163 @@ Status code     | Description
...
+User management endpoints are released in Beta and may change depending on feedback during trial.
+</aside>
+
+Enable administrators (users with the `Manage Role` permission) to remove a user from a specified group or organisation
+<aside class="notice">
+To remove a user from a user group, the following conditions must be satisfied:
+
+1. The requesting user cannot remove themselves
+2. Only users with the `Manage Role` permission can remove other users
+3. Administrators (users with `Manage Role` permission) can remove other administrator users from the organisation
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: from any group in the organisation

File: source/index.html.md [hunk 26]
```diff
@@ -2639,3 +2639,163 @@ Status code     | Description
...
+4. If a user is added to an administrator group (group with `Manage Role` permission) within the organisation, the user gets the administrator (`Manage Role` permission) privileges,
+   if the user is the only member of a group and we try to remove the user, a check is made to see if the organisation has more than one administrators, if yes, the user can be removed from the group.
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Is this really happening? This code ` if (roleDoc.members.length > 1) {
            return true;
          }` seems to bypass this condition

File: source/index.html.md [hunk 27]
```diff
@@ -2639,3 +2639,163 @@ Status code     | Description
...
+5. If a user is added to an administrator group (group with `Manage Role` permission), the user inherits the administrator (`Manage Role` permission) privileges,
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Is this the same as 4. above? Sounds like you can merge 4. and 5. I.e. say 

If the user to be removed belongs to a group with `Manage Role` permission they inherit the `Manage Role` permission. In that case the user can only be removed from the group if they are not the only administrator in the organisation. After removal from the group the user loses the previously inherited `Manage Role` permission.

File: source/index.html.md [hunk 28]
```diff
@@ -2639,3 +2639,163 @@ Status code     | Description
...
+   if the user is removed from this group, he will lose his adminstrator (`Manage Role` permission) privileges.
+6. The last remaining administrator of the organisation cannot be removed
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: The last remaining user with `Manage Role` permission in the organisation cannot be removed

File: source/index.html.md [hunk 29]
```diff
@@ -2725,9 +2725,7 @@ User Accounts      | Description
 `firstName`        | The first name of the user (with maximum character limit of 150)
 `lastName`         | The last name of the user (with maximum character limit of 150)
 `email`            | Email Address of the user (with maximum character limit of 200)
-`password`         | Password of the user (with maximum character limit of 200 and minimum character limit of 6) (optional)
-`hashedPassword`   | Hashed password of the user (with maximum character limit of 60) (optional)
-
+`password`         | Password of the user (with maximum character limit of 200 and minimum character limit of 6) (optional)
```
Comment (dimitrikoubaroulis) [inline source/index.html.md]: Here we should describe what will happen if the user does not provide a password for the added user

File: source/index.html.md (hunk 26 above)
Comment (mrinalirao) [reply source/index.html.md]: This is only if the group is admin group..if the group is admin group..it checks this condition:
if (roleDoc.members.length > 1)  and returns true. However if it is not the admin group..it goes down further to count the total admins in the org
``` soter.getRole orgID, (error, orgDoc) ->
        if error? then return callback error
        if orgDoc.role_type != "org"
          request.log ['error'], "Expected organisation doc but got group doc"
          return callback "Expected organisation doc but got group doc"
        countAdmins orgDoc, soter, (err, res) ->```

**Pull request merged**